managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerImpl.java

```diff
@@ -2513,7 +2531,25 @@ void internalTrimLedgers(boolean isTruncate, CompletableFuture<?> promise) {
             advanceCursorsIfNecessary(ledgersToDelete);
 
             PositionImpl currentLastConfirmedEntry = lastConfirmedEntry;
+
             // Update metadata
+            // Mark deletable ledgers
+            Set<Long> deletableLedgers = Stream
+                    .concat(
+                            ledgersToDelete.stream()
+                                    .filter(ls -> !ls.getOffloadContext().getBookkeeperDeleted()),
+                            offloadedLedgersToDelete.stream())
+                    .map(LedgerInfo::getLedgerId)
+                    .collect(Collectors.toSet());
+
+            // Mark deletable offloaded ledgers
+            Set<Long> deletableOffloadedLedgers = ledgersToDelete.stream()
+                    .filter(ls -> ls.getOffloadContext().hasUidMsb())
```

Comment:
You'd better check `ls.getOffloadContext().isComplete` instead of `ls.getOffloadContext().hasUidMsb()`

Reply:
This is mainly implemented according to the original logic. See https://github.com/apache/pulsar/blob/c18d64526708c6dd051217feba3c31a73507620c/managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerImpl.java#L2716-L2729